### PR TITLE
Explicit Imports

### PR DESCRIPTION
**Pull request by Borda**
make the import explicit so it is easier to navigate from where each function originate...

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced YOLOv5 codebase with refined imports, utility functions, and module usage.

### 📊 Key Changes
- 🧹 Cleaned up import statements across various files, removing unnecessary imports.
- 🛠️ Modularized code by replacing broad import statements with specific ones.
- 🚀 Improved code structure and readability by using functions directly from their modules instead of through utility wrappers.
- 🏗️ Refactored Google utils usage: replaced `google_utils.attempt_download` with `attempt_download`.
- 🔩 Various refactors for code simplification including model initialization, utility function usage, and error handling.
- 💡 Simplified and reorganized code by removing redundant utility calls and batch norm fusion in several models.
- ✨ Integrated new logic and methods for model training, testing, and exporting with updated utilities.

### 🎯 Purpose & Impact
- 🚀 Enhances maintainability and understandability of the code by making imports and utility functions more explicit.
- 💡 Multiple small refactors contribute to a clearer flow of logic, potentially reducing errors and making the codebase more approachable for contributors.
- 👁️‍🗨️ Impacts current and future developers, easing the development process with a more organized and structured codebase.
- 🌱 No direct impact on end-user functionality, but indirectly leads to a more stable and reliable codebase.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@Borda thank you for the PR! Yes I saw this aspect of the updates in your pip PR. I don't come from a software engineering background (physics and aerospace engineering), so I'm flying blind when it comes to software best practices.

Is this best practices in python coding to explicitly import functions rather than relying on * or importing the entire package?

EDIT: My main worry is the increase in the codebase, especially in the main files train.py, test.py.

**Comment by Borda** [comment]
> @Borda thank you for the PR! Yes I saw this aspect of the updates in your pip PR. I don't come from a software engineering background (physics and aerospace engineering), so I'm flying blind when it comes to software best practices.

sure, happy to help

> Is this best practices in python coding to explicitly import functions rather than relying on * or importing the entire package?

yes it is, as using `*` is black magic and also it may import plant of function you do not need...

**Comment by glenn-jocher** [comment]
@Borda ok got it. Well I suppose we should do this then!

First we have to wait for @MagicFrogSJTU to complete a bit of a foundational rewrite of test.py, he said he'd have this completed by monday and I advised that I wouldn't modify the files in the meantime. There is not a PR currently open for this, but I assume it will arrive over the weekend. After that there are a few other items such as the logging PR and then I'll try to get this in after that!

**Comment by Borda** [comment]
> @Borda ok got it. Well I suppose we should do this then!

one more reason why not use * is that if you for some reason change import in any chained imported files you may be missing some function and then you spend some time on debugging what is wrong even you did not touch the failing file...

> First we have to wait for @MagicFrogSJTU to complete a bit of a foundational rewrite of test.py

 would say that it is not needed to wait, it is just about imports so with this change he just continues using explicit importing... :]

**Comment by Borda** [comment]
@glenn-jocher then let me know when it is actual and I will rebase/resolve conflicts... :]

**Comment by Borda** [comment]
@glenn-jocher resolved now... :rabbit:

**Comment by glenn-jocher** [comment]
@Borda thanks for the rebase! I ran the checks again manually today and saw failures. I also saw the same result in https://github.com/ultralytics/yolov5/pull/519. I think this is unrelated to the PRs, it's possible the github actions python version incremented to 3.8.3, as I see some simple python errors in the code now. I'm going to investigate a bit.

**Comment by glenn-jocher** [comment]
@Borda nevermind, I see pytorch 1.6.0 was released today, surely not a coincidence. It looks like some compatibility updates are in order on my part...

**Comment by Borda** [comment]
> @Borda thanks for the rebase! I ran the checks again manually today and saw failures. I also saw the same result in #519. I think this is unrelated to the PRs, it's possible the github actions python version incremented to 3.8.3, as I see some simple python errors in the code now. I'm going to investigate a bit.

it may be possible, for that you could freeze your python version to 3.8.2 to whatever is you preference

> @Borda nevermind, I see pytorch 1.6.0 was released today, surely not a coincidence. It looks like some compatibility updates are in order on my part...

you may add PT version in your CI builds like specify 1,5 and/or 1.6

**Comment by Borda** [comment]
@glenn-jocher how about this one? :]

**Comment by Borda** [comment]
@glenn-jocher is there an interest in fixing imports? otherwise, I'll close this one...

**Comment by glenn-jocher** [comment]
@Borda yes, let's do it. I'll see if I can resolve the conflicts.

**Comment by Borda** [comment]
> Borda yes, let's do it. I'll see if I can resolve the conflicts.

resolved :]

**Comment by glenn-jocher** [comment]
Oh great! That's odd, I see one more conflict. Let me try the new rebase command. It didn't work before on a different PR, I'll try it here.

/rebase

**Comment by Borda** [comment]
> Oh great! That's odd, I see one more conflict. Let me try the new rebase command. It didn't work before on a different PR, I'll try it here.
> 
> /rebase

that rebase usually does not work...

**Comment by glenn-jocher** [comment]
Awesome, thank for your hard work!!
